### PR TITLE
BERT Inference on CPU

### PR DESCRIPTION
**Pull request by mfuntowicz**
HTML is just for rendering the readable content for review. It will be removed before merging.

### CODE REVIEW

File: benchmark-2021-part1.md [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+Limiting the number of resources involved in computation is done by limiting the number of threads involved in 
+**intra** operations (_**intra** here means inside an operator doing computation, also known as "kernel"_). 
+
+This is achieved through the following API:
+- PyTorch: `torch.set_num_threads(x)`
+- TensorFlow: `tf.config.threading.set_intra_op_parallelism_threads(x)`
+
+![pytorch_tensorflow_intraops_scaling](assets/19_benchmark_2021_part1/imgs/core_count_scaling.svg)
+
+As you can see, depending on the problem size, the number of threads involved in the computations has a positive impact
```
Comment (n1t0) [inline benchmark-2021-part1.md:143]: I find the phrasing a bit weird here, what about something like this:

> As you can see, increasing the number of threads usually has a positive impact, but the size of the problem limits this impact a lot.
>
> For small-sized problems, using only 8 or 16 threads already gives the best performances. The same applies to medium-sized problems where using a single CPU (24 threads) gives the best results. For large-sized problems though, the overhead of the cross-socket communication is covered by the computations cost, thus benefiting from using all the cores available on the system.

File: benchmark-2021-part1.md [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+computations for both PyTorch and TensorFlow.
+
+Limiting the number of resources involved in computation is done by limiting the number of threads involved in 
+**intra** operations (_**intra** here means inside an operator doing computation, also known as "kernel"_). 
+
+This is achieved through the following API:
+- PyTorch: `torch.set_num_threads(x)`
+- TensorFlow: `tf.config.threading.set_intra_op_parallelism_threads(x)`
+
+![pytorch_tensorflow_intraops_scaling](assets/19_benchmark_2021_part1/imgs/core_count_scaling.svg)
```
Comment (n1t0) [inline benchmark-2021-part1.md:141]: I find the title a bit misleading here. I was expecting the batch size to change along the way but it seems to be always 1. Also, I find it a bit hard to parse the various parameters.

File: benchmark-2021-part1.md [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+<!-- The figure below explains the above visually ( [source](https://appuals.com/how-does-hyper-threading-work-in-intel-core-i7-processors/) ) -->
+
+
+<img class="centered" alt="Intel Hyper Threading technology" src="assets/19_benchmark_2021_part1/imgs/hyper_threading_explained.png" />
+
+
+Back to our model inference workload... If you think about it, in a perfect world and very optimized setup, computations take the majority of time. 
+
+In this context, using the logical cores shouldn't bring us any performance benefits because our computation work
+(*task A in the example above*) is not letting the second one (*task B*) being schedule on the CPU core.
```
Comment (n1t0) [inline benchmark-2021-part1.md:177]:
```suggestion
(*task A in the example above*) is not letting the second one (*task B*) being scheduled on the CPU core.
```

File: benchmark-2021-part1.md [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+On our way towards optimizing CPU inference for better usage of the CPU cores you might have already seen -_at least for the
+past 20 years_- modern CPUs specifications reports "cores" and "threads" or "physical" and "logical" numbers. 
+These notions refer to a mechanism called **Simultaneous Multi-Threading** (SMT).
+
+To put things simple, imagine two tasks **A** and **B**, executing in parallel, each on its own thread.  
+At some point, there is a high probability these two tasks will have to wait for some resources to be fetched from main memory, SSD, HDD 
+or even the network.  
+During these periods the core executing the task is in an **Idle** state waiting for the resources to arrive, and effectively doing nothing...
+
+Now, with **SMT**, the **two threads for task A and B** will be scheduled on the same **physical core**,
```
Comment (n1t0) [inline benchmark-2021-part1.md:163]: Might be worth mentioning that `Hyper Threading` is the name Intel gave to `SMT`.

File: benchmark-2021-part1.md [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+our inference as linearly as possible.
+
+
+Let's start simple, if we want to spawn 2 instances, one on each socket with 24 cores assigned:
+```shell
+numactl -C 0-23 -m 0 python3 src/main.py model=bert-base-cased batch_size=1 sequence_length=128 backend.name=pytorch backend.num_threads=24
+numactl -C 24-47 -m 1 python3 src/main.py model=bert-base-cased batch_size=1 sequence_length=128 backend.name=pytorch backend.num_threads=24
+```
+
+Starting from here, both instances doesn't share any resources between them and everything is operating as its maximum efficiency from a
```
Comment (n1t0) [inline benchmark-2021-part1.md:296]:
```suggestion
Starting from here, both instances don't share any resources between them and everything is operating at its maximum efficiency from a 
```

File: benchmark-2021-part1.md [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+
+The following section summarizes the performances of Multi-Stream Inference, leveraging all the knobs explained above.
+
+### Batch size scaling results
+
+The first chart below reports the best latency setup depending on the problem size (_w.r.t the sequence length_).
+This corresponds to taking the **maximum** inference time for all the instance for a same problem size.
+The second one reports the actually scaling efficiency by **summing** the throughput for all instances for a same problem size.
+
+![Batch scaling experiment for PyTorch and Tensorflow](assets/19_benchmark_2021_part1/imgs/batch_scaling_exp.svg)
```
Comment (n1t0) [inline benchmark-2021-part1.md:336]: I find the parameters hard to read on this graph too.

File: benchmark-2021-part1.md [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+
+![Batch scaling experiment for PyTorch and Tensorflow](assets/19_benchmark_2021_part1/imgs/batch_scaling_exp_throughput.svg)
+
+## Conclusion
+
+Through this blog post we covered the basic BERT inference results for both PyTorch and TensorFlow one can expect without from a simple PyPi install and without any further tuning.
```
Comment (n1t0) [inline benchmark-2021-part1.md:342]:
```suggestion
Through this blog post, we covered the basic BERT inference results for both PyTorch and TensorFlow one can expect from a simple PyPi install and without any further tuning.
```

**Review by n1t0 — COMMENTED** [review]
Nice one @mfuntowicz! I left a few comments about some typos and things I had trouble understanding.

File: benchmark-2021-part1.md [hunk 8]
```diff
@@ -0,0 +1,431 @@
...
+```
+
+Then we specify the core and memory affinity through `numactl` spawning all (_and only_) the **physical** cores (_Right_):
+```shell
+numactl -C 0-47 -m 0,1 python3 src/main.py model=bert-base-cased backend.name=pytorch batch_size=1 sequence_length=128
+```
+
+<img class="centered" alt="htop CPU usage without and with numactl process affinity set" src="assets/19_benchmark_2021_part1/imgs/numa_combined.svg" /> 
+
+As you can see, without any specific tuning, PyTorch and TensorFlow dispatch the work on a single core, using both physical **and** logical cores.
```
Comment (lhoestq) [inline benchmark-2021-part1.md:259]:
```suggestion
As you can see, without any specific tuning, PyTorch and TensorFlow dispatch the work on a single CPU, using both physical **and** logical cores.  
```
Shouldn't be it "CPU" or "processor" here instead of "core" ? This sentence is confusing

File: benchmark-2021-part1.md [hunk 9]
```diff
@@ -0,0 +1,489 @@
...
+
+<div class="author-card">
+    <a href="/mfuntowicz">
+        <img class="avatar avatar-user" src="https://aeiljuispo.cloudimg.io/v7/https://s3.amazonaws.com/moonup/production/uploads/1583858935715-5e67c47c100906368940747e.jpeg?w=200&h=200&f=face" title="Gravatar">
+        <div class="bfc">
+            <code>mfuntowicz</code>
+            <span class="fullname">Morgan Funtowicz</span>
+        </div>
+    </a>
+</div>
```
Comment (julien-c) [inline benchmark-2021-part1.md:27]: For consistency, can you add the customary button/link to edit the blogpost on GitHub, like on other blogposts? @mfuntowicz
